src/qn_model.jl

```diff
@@ -0,0 +1,64 @@
+export QuasiNewtonModel, LBFGSModel, LSR1Model,
+       reset!,
+       obj, grad, grad!,
+       cons, cons!, jac_coord, jac, jprod, jprod!, jtprod, jtprod!,
+       hprod, hess_op,
+       push!
+
+abstract QuasiNewtonModel <: AbstractNLPModel
+
+type LBFGSModel <: QuasiNewtonModel
+  meta :: NLPModelMeta
+  model :: AbstractNLPModel
+  op :: LBFGSOperator
+end
+
+type LSR1Model <: QuasiNewtonModel
+  meta :: NLPModelMeta
+  model :: AbstractNLPModel
+  op :: LSR1Operator
+end
+
+"Construct a `LBFGSModel` from another type of model."
+function LBFGSModel(nlp :: AbstractNLPModel; memory :: Int=5)
+  op = LBFGSOperator(nlp.meta.nvar, memory)
+  return LBFGSModel(nlp.meta, nlp, op)
+end
+
+"Construct a `LSR1Model` from another type of nlp."
+function LSR1Model(nlp :: AbstractNLPModel; memory :: Int=5)
+  op = LSR1Operator(nlp.meta.nvar, memory)
+  return LSR1Model(nlp.meta, nlp, op)
+end
+
+# retrieve counters from underlying model
+for counter in fieldnames(Counters)
+  @eval begin
+    $counter(nlp :: QuasiNewtonModel) = $counter(nlp.model)
+    export $counter
+  end
+end
+
+function reset!(nlp :: QuasiNewtonModel)
+  reset!(nlp.model.counters)
+  reset!(nlp.op)
```

Comment:
This is mysterious to me. There's a `using LinearOperators` at the top of `NLPModels.jl` (which `include`s `qn_model.jl`). Why doesn't multiple dispatch do its job here?